providers/common/ciphers/aes_ccm_s390x.inc

```diff
@@ -0,0 +1,263 @@
+    }
+    l = flags & 0x7;
+    ctx->ccm.s390x.nonce.b[0] = l;
+
+    /*-
+     * Reconstruct length from encoded length field
+     * and initialize it with counter value.
+     */
+    n = 0;
+    for (i = 15 - l; i < 15; i++) {
+        n |= ctx->ccm.s390x.nonce.b[i];
+        ctx->ccm.s390x.nonce.b[i] = 0;
+        n <<= 8;
+    }
+    n |= ctx->ccm.s390x.nonce.b[15];
+    ctx->ccm.s390x.nonce.b[15] = 1;
+
+    if (n != len)
+        return 0;      /* length mismatch */
+
+    if (enc) {
+        /* Two operations per block plus one for tag encryption */
+        ctx->ccm.s390x.blocks += (((len + 15) >> 4) << 1) + 1;
+        if (ctx->ccm.s390x.blocks > (1ULL << 61))
+            return 0;      /* too much data */
+    }
+
+    num = 0;
+    rem = len & 0xf;
+    len &= ~(size_t)0xf;
+
+    if (enc) {
+        /* mac-then-encrypt */
+        if (len)
+            s390x_kmac(in, len, ctx->ccm.s390x.fc,
+                       &ctx->ccm.s390x.kmac);
+        if (rem) {
+            for (i = 0; i < rem; i++)
+                ctx->ccm.s390x.kmac.icv.b[i] ^= in[len + i];
+
+            s390x_km(ctx->ccm.s390x.kmac.icv.b, 16,
+                     ctx->ccm.s390x.kmac.icv.b,
+                     ctx->ccm.s390x.fc, ctx->ccm.s390x.kmac.k);
+        }
+
+        CRYPTO_ctr128_encrypt_ctr32(in, out, len + rem, &ctx->ccm.ks.ks,
+                                    ctx->ccm.s390x.nonce.b, ctx->ccm.s390x.buf.b,
+                                    &num, (ctr128_f)AES_ctr32_encrypt);
+    } else {
+        /* decrypt-then-mac */
+        CRYPTO_ctr128_encrypt_ctr32(in, out, len + rem, &ctx->ccm.ks.ks,
+                                    ctx->ccm.s390x.nonce.b, ctx->ccm.s390x.buf.b,
+                                    &num, (ctr128_f)AES_ctr32_encrypt);
+
+        if (len)
+            s390x_kmac(out, len, ctx->ccm.s390x.fc,
+                       &ctx->ccm.s390x.kmac);
+        if (rem) {
+            for (i = 0; i < rem; i++)
+                ctx->ccm.s390x.kmac.icv.b[i] ^= out[len + i];
+
+            s390x_km(ctx->ccm.s390x.kmac.icv.b, 16,
+                     ctx->ccm.s390x.kmac.icv.b,
+                     ctx->ccm.s390x.fc, ctx->ccm.s390x.kmac.k);
+        }
+    }
+    /* encrypt tag */
+    for (i = 15 - l; i < 16; i++)
+        ctx->ccm.s390x.nonce.b[i] = 0;
+
+    s390x_km(ctx->ccm.s390x.nonce.b, 16, ctx->ccm.s390x.buf.b,
+             ctx->ccm.s390x.fc, ctx->ccm.s390x.kmac.k);
+    ctx->ccm.s390x.kmac.icv.g[0] ^= ctx->ccm.s390x.buf.g[0];
+    ctx->ccm.s390x.kmac.icv.g[1] ^= ctx->ccm.s390x.buf.g[1];
+
+    ctx->ccm.s390x.nonce.b[0] = flags;    /* restore flags field */
+    return 1;
+}
+
+static int s390x_aes_ccm_auth_encrypt(PROV_AES_CCM_KEY *ctx,
+                                      const unsigned char *in,
+                                      unsigned char *out, size_t len,
+                                      unsigned char *tag, size_t taglen)
+{
+    return s390x_aes_ccm_auth_encdec(ctx, in, out, len, 1);
```

Comment:
In case tag != NULL, the tag value has to be copied. ccm test cases dont find this, they probably fetch it using gettag routine. Though, the test cases for tls code path find it.